src/validation.cpp

```diff
@@ -2245,6 +2248,173 @@ bool BitcoindRPCCheck(const bool init)
+        return false;
+    }
+
+    // Txid must be included
+    if (stack[0].size() != 32) {
+        return false;
+    }
+    uint256 txid(stack[0]);
+
+    // Get output number
+    int nOut = CScriptNum(stack[1], true).getint();
+
+    // Get output value. Special 8 byte length to capture possible bitcoin values
+    CAmount value = CScriptNum(stack[2], true, 8).getint64();
+
+    // Get asset type
+    if (stack[3].size() != 32) {
+        return false;
+    }
+    CAsset asset(stack[3]);
+
+    // Get genesis blockhash
+    if (stack[4].size() != 32) {
+        return false;
+    }
+    uint256 gen_hash(stack[4]);
+
+    // Get witness program
+    CScript witness_program(stack[5].begin(), stack[5].end());
+    int version = -1;
+    std::vector<unsigned char> witnessProgram;
+    if (!witness_program.IsWitnessProgram(version, witnessProgram)) {
+        return false;
+    }
+
+    // Get serialized transaction
+    Sidechain::Bitcoin::CTransactionRef pegtx;
+	try {
+		CDataStream pegtx_stream(stack[6], SER_NETWORK, PROTOCOL_VERSION);
+        pegtx_stream >> pegtx;
+        if (!pegtx_stream.empty()) {
+	        return false;
+		}
+	} catch (std::exception& e) {
+        // Invalid encoding of transaction
+        return false;
+    }
+
+    // Get txout proof
+    Sidechain::Bitcoin::CMerkleBlock merkle_block;
+    std::vector<uint256> txHashes;
+    std::vector<unsigned int> txIndices;
+
+	try {
+        CDataStream merkleBlockStream(stack[7], SER_NETWORK, PROTOCOL_VERSION);
+        merkleBlockStream >> merkle_block;
+        if (!merkleBlockStream.empty() || !CheckBitcoinProof(merkle_block.header.GetHash(), merkle_block.header.nBits)) {
+           return false;
+        }
+        if (merkle_block.txn.ExtractMatches(txHashes, txIndices) != merkle_block.header.hashMerkleRoot || txHashes.size() != 1) {
+            return false;
+        }
+	} catch (std::exception& e) {
+        // Invalid encoding of merkle block
+        return false;
+    }
+
+    // Check that transaction matches txid
+    if (pegtx->GetHash() != txid) {
+        return false;
+    }
+
+    // Check that the merkle proof corresponds to the txid
+    if (txid != txHashes[0]) {
+        return false;
+    }
+
+    // Check the genesis block corresponds to a valid peg (only one for now)
+    if (gen_hash != Params().ParentGenesisBlockHash()) {
+        return false;
+    }
+
+    // Check the asset type corresponds to a valid pegged asset (only one for now)
+    if (asset != Params().GetConsensus().pegged_asset) {
+        return false;
+    }
+
+    // Check the transaction nout/value matches
+    if (nOut < 0 || (unsigned int)nOut >= pegtx->vout.size() || value != pegtx->vout[nOut].nValue) {
+        return false;
+    }
+
+    // Check that the witness program matches the p2ch on the transaction output
+    opcodetype opcodeTmp;
+    unsigned char tweak[32];
+    CSHA256().Write(witness_program.data(), witness_program.size()).Finalize(tweak);
```

Comment:
The tweak should somehow include the fedpegScript. In the original scheme each pubkey goes into the hash that computes its tweak, which is the standard way to do pay-to-contract and is always secure.

As written, it is possible in principle for the functionaries to adversarially choose their keys to steal some existing Bitcoin output. (They'd be unable to process any other peg-outs, and the existing output would need to be in the correct multisig form, so this is a silly attack scenario, but it is there.)

So there are two options: (1) include `fedpegScript` in its entirety in the tweak, or (2) in the key-tweaking loop recompute the tweak each time using the standard pay-to-contract scheme.